src/sqlfluff/core/rules/base.py

```diff
@@ -416,6 +435,59 @@ def raw_segments(self):
         )
 
 
+class CrawlBehavior:
```

Comment:
I created this class by moving the parse tree traversal from `crawl()` to here. Making it a separate class makes it easier to vary the behavior, make it configurable, etc. with less risk of breaking other things (or having future changes to the other parts of the linter loop break _this_ code).